docs/dev/allocation.qmd

```diff
@@ -0,0 +1,279 @@
+### Setting up the objective function
+The optimization objective is the sum of three quadratic error terms.
+The quadratic terms are defined with `add_objective_term` function.
+
+In `set_objective_priority`, it sets the objective function for a given priority with the following steps:
+
+- First, it treats the subnetworks as user demand nodes and adds the quadratic terms of the main network.
+- Then it loops over all the edges in allocation.
+- Based on the type of the node that the edge is pointing to (user demand or flow demand), it adds the corresponding quadratic terms.
+- Finally, it does the same to the edges that start from a level demand node.
+
+### Setting the constraints and capacities
+In the function `set_initial_values`, the following capacities and demands are initialized:
+
+- Source capacities come from the physical layer
+- Edge capacities derived from the maximum capacities between the connected nodes
+- Basin capacities
+- Buffer capacities
+- User demands fractional return flow
+- Demands either come from the Ribasim model or are set via the BMI
+
+As shown below, these functions set the capacities to the corresponding initial values.
+
+```julia
+set_initial_capacities_source!(allocation_model, p)
+set_initial_capacities_edge!(allocation_model, p)
+set_initial_capacities_basin!(allocation_model, p, u, t)
+set_initial_capacities_buffer!(allocation_model)
+set_initial_capacities_returnflow!(allocation_model)
+
+set_initial_demands_user!(allocation_model, p, t)
+set_initial_demands_level!(allocation_model, u, p, t)
+set_initial_demands_flow!(allocation_model, p, t)
+```
+
+These capacities determine the constraints of the optimization problem.
+Take `set_initial_capacities_source` as an example, the right-hand-side values of the `source_constraints` are set to the `source_capacity`.
+
+```julia
+for edge_metadata in values(graph.edge_data)
+    (; edge) = edge_metadata
+    if graph[edge...].subnetwork_id_source == subnetwork_id
+        # If it is a source edge for this allocation problem
+        if edge ∉ main_network_source_edges
+            # Reset the source to the averaged flow over the last allocation period
+            source_capacity = mean_input_flows[edge][]
+            JuMP.set_normalized_rhs(
+                source_constraints[edge],
+                # It is assumed that the allocation procedure does not have to be differentiated.
+                source_capacity,
+            )
+        end
+    end
+end
+```
+
+Apart from the `set_initial_` function above, capacities of `inlet` are the allocated capacities from the main network to the subnetworks.
+Source constraints will be adapted based on the optimization type.
+This function is called separately and thus not part of the `set_initial_values`.
+
+## Looping over priorities
+### Updating capacities
+While optimizing a given priority, the function `set_capacities_flow_demand_outflow` updates the constraints `flow_demand_outflow`.
```

Comment:
Some additional information about the buffer for nodes with a flow demand:
- At priorities where there is no flow demand, flow can go freely trough the node. When there is flow demand, flow is directed into the buffer. This is to make sure that flow can go to the node with the flow demand, even though the flow might have nowhere to go after that node.